### PR TITLE
feat: upgrade `@asyncapi/parser` to v2.0.0 and use the new parser-api 

### PR DESCRIPTION
**Pull request by smoya**
Replaces https://github.com/asyncapi/generator/pull/833

**Description**

Use Parser-JS v2, and new Custom Schema Parsers:

- add new template config field apiVersion.
- update relevant source code to handle new and old API
- add some utils
- update tests

**Related issue(s)**
Resolves https://github.com/asyncapi/generator/issues/825
Part of https://github.com/asyncapi/parser-js/issues/481

### CODE REVIEW

**Review by fmvilas — DISMISSED** [review]

**Comment by jonaslagoni** [comment]
This is a breaking version change, no?

**Comment by smoya** [comment]
> This is a breaking version change, no? Or is the API 100% backward compatible with the old parser api?

It is compatible. Default value for `apiVersion` field is to use `1` that means old API.
But now that you mention this, I think it would make more sense to map this field to the version of the [Parser-API](https://github.com/asyncapi/parser-api/releases) the template uses and is compatible with.
That means no field for old API (Parser-JS v1.x) and `v1` for `v1.x` of the Parser-API.

Do you think it makes sense? cc @magicmatatjahu @jonaslagoni @derberg @fmvilas

**Comment by magicmatatjahu** [comment]
`apiVersion` is made so that there is no breaking changes in templates and people have time to rewrite to api `1.x`. We can change the value from `1` to `0` and it will mean the old api, but it should still be the default value on the `1.x` generator. In `2.x` generator should use `1.x` API.

**Comment by smoya** [comment]
> `apiVersion` is made so that there is no breaking changes in templates and people have time to rewrite to api `1.x`. We can change the value from `1` to `0` and it will mean the old api, but it should still be the default value on the `1.x` generator. In `2.x` generator should use `1.x` API.

But by API we mean Parser-API in the end right? So we could say 0 for old api, 1 for v1 Parser-API, and so on?

**Comment by magicmatatjahu** [comment]
@smoya Yes :)

**Comment by smoya** [comment]
> > `apiVersion` is made so that there is no breaking changes in templates and people have time to rewrite to api `1.x`. We can change the value from `1` to `0` and it will mean the old api, but it should still be the default value on the `1.x` generator. In `2.x` generator should use `1.x` API.
> 
> But by API we mean Parser-API in the end right? So we could say 0 for old api, 1 for v1 Parser-API, and so on?

I think better UX will be:

- Field not present: Use old [Parser-JS v1 API](https://github.com/asyncapi/parser-js/tree/v1.18.1#api-documentation).
- Field is present: Use the specified version of the [Parser-API](https://github.com/asyncapi/parser-api) (if supported, error otherwise).

**Comment by fmvilas** [comment]
I think that change makes sense to ease the migration process but let's take into account that making `v2` as the default in the future will be another major version of the Generator. Which is fine, just want to make sure we're all aware.

**Review by fmvilas — DISMISSED** [review]

**Comment by smoya** [comment]
> I think that change makes sense to ease the migration process but let's take into account that making `v2` as the default in the future will be another major version of the Generator. Which is fine, just want to make sure we're all aware.

Yes. This step is only for keeping backward compatibility. That's the reason it won't be released as a major version but just as a minor. 
However, the next Generator major version (`v2`) could keep the same behavior of using the old API if `apiVersion` field is empty if we want. 
Let me explain this: The support of defaulting to the old API can be kept until Parser-JS drops support for it, which is not really planned to happen yet (even though we should eventually do it).

Another thing to clarify, is that a major version release of the Generator doesn't need to be just because a new major version of the Parser-API is supported. It might happen that we need to do a major release because another breaking change not related to such API. 

No matters what, but we should specify in the README always which versions of the Parser-API we support on the Generator and what is the default in case of missing `apiVersion`.

**Comment by smoya** [comment]
At this point, I wonder if we should remove, in version 2.x of the generator, the field `generator` in the config, which states as:

> A string representing the generator version-range the template is compatible with. This value must follow the [semver](https://nodejs.dev/learn/semantic-versioning-using-npm) syntax. E.g., >=1.0.0, >=1.0.0 <=2.0.0, ~1.0.0, ^1.0.0, 1.0.0, etc. [Read more about semver](https://docs.npmjs.com/about-semantic-versioning).

It still might be useful for templates that require extra functionality not implemented in old Generator versions.

Not sure about that, just want to raise awareness @magicmatatjahu @fmvilas

**Comment by magicmatatjahu** [comment]
@smoya As I remember it was created only for changes in the generator, e.g. your generator uses the old version of generator, because new functionality "breaks" your output from generator - by this user of template know about limitation and you as developer of template has a time to add support for new functionalities. I would leave that :)

cc @derberg

File: docs/configuration-file.md
Comment (smoya) [reply docs/configuration-file.md:11]: Good point. I can add that, because I consider is how the generator should behave so it doesn't goes with unexpected behaviors. However, I'm wondering if that will be considered as breaking change: templates that, for some reason, include now the `apiVersion` field (I know it wasn't declared, but imagine someone added it because "whatever") will start receiving those exceptions.
Obviously, this is not gonna happen with the list of official supported templates. 

In my opinion, it is not a BC because `apiVersion` was never supported nor even mentioned in the documentation until this PR. WDYT?

**Review by derberg — DISMISSED** [review]
yes, `generator` config prop is needed. Basically imagine we do what @fmvilas mentioned in the other comment, that we make v2 default and release 2.0, without `generator` prop in the template, template users instead of clear info "that template is not compatible with 2.0" they will get some parser errors.

great work, probably https://github.com/asyncapi/generator/blob/master/test/__mocks__/%40asyncapi/parser.js can be removed

File: docs/configuration-file.md
Comment (smoya) [reply docs/configuration-file.md:11]: I added this change so it throws an exception during parameters validation if the `apiVersion` is present but not valid value (within a range).

See https://github.com/asyncapi/generator/pull/960/commits/dbb27c469f174dff0f42be29b5b8fa22d0b6ac7e

cc @derberg @fmvilas

**Review by derberg — APPROVED** [review]
LGTM 🚀 
If there are no new comments before Wednesday feel free to merge on Wednesday

**Comment by derberg** [comment]
@Florence-Njeri @pratik2315 please have a look on docs changes, your approval is required

**Comment by smoya** [comment]
No comments from last time, so I'm gonna merge this

**Comment by derberg** [comment]
@smoya you can't merge until @Florence-Njeri or @pratik2315 approves

**Comment by smoya** [comment]
> @smoya you can't merge until @Florence-Njeri or @pratik2315 approves

Yeah, I feel like an adventurer today 😅

File: docs/configuration-file.md
Comment (derberg) [reply docs/configuration-file.md:11]: oh yeah, good catch!

@smoya

Comment (smoya) [reply docs/configuration-file.md:11]: Parser-API is not the same as the API of Parser-JS v1. Links are not the same either.

Comment (derberg) [reply docs/configuration-file.md:11]: but you have it here already `Determines which **major** version of the [Parser-API](https://github.com/asyncapi/parser-api)`

**Review by Florence-Njeri — APPROVED** [review]

**Review by derberg — APPROVED** [review]

**Pull request merged**
